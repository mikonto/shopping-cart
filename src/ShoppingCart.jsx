import styled from "styled-components";

const StyledShoppingCart = styled.div`
  background: white;
  flex-grow: 1;
  padding: 4px;
`;

const ShoppingCart = () => {
  return (
    <>
      <StyledShoppingCart>Shopping Cart</StyledShoppingCart>
    </>
  );
};

export default ShoppingCart;
