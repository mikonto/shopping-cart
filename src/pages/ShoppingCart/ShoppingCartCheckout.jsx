import styled from "styled-components";
import PropTypes from "prop-types";

const StyledShoppingCartCheckout = styled.div`
  background: white;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 12px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  border-radius: 4px;
`;

const StyledUl = styled.ul`
  margin: 0;
`;

const StyledLi = styled.li`
  margin: 6px;
`;

const NoBulletLi = styled.li`
  margin: 6px;
  list-style-type: none;
`;

const Separator = styled.div`
  height: 1px;
  width: calc(100% - 40px);
  background-color: #c0c0c0;
  margin: 20px;
`;

const BlueButton = styled.button`
  margin-top: auto;
  margin-left: auto;
  background-color: skyblue;
  padding: 8px 12px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.3s ease;

  &:hover {
    background-color: #6ca5bc;
  }
`;

const ShoppingCartCheckout = ({ shoppingCart }) => {
  const subTotal = () => {
    const sum = shoppingCart.reduce((total, cartItem) => {
      return total + cartItem.details.price * cartItem.quantity;
    }, 0);

    return Number(sum.toFixed(2));
  };

  const totalSum = () => {
    const sum = subTotal() + 8;
    return Number(sum.toFixed(2));
  };

  return (
    <StyledShoppingCartCheckout>
      <p>
        <b>Order summary</b>
      </p>
      <StyledUl>
        {shoppingCart.map((cartItem) => (
          <StyledLi key={cartItem.details.id}>
            {cartItem.quantity} x {cartItem.details.title}
          </StyledLi>
        ))}
      </StyledUl>
      <Separator />
      <StyledUl>
        <NoBulletLi>Products total: {subTotal()}€</NoBulletLi>
        <NoBulletLi>Shipping costs: 8€</NoBulletLi>
      </StyledUl>
      <Separator />
      <StyledUl>
        <NoBulletLi>
          <b>Total: {totalSum()}€</b>
        </NoBulletLi>
      </StyledUl>
      <BlueButton>To checkout</BlueButton>
    </StyledShoppingCartCheckout>
  );
};

ShoppingCartCheckout.propTypes = {
  shoppingCart: PropTypes.array,
};

export default ShoppingCartCheckout;
