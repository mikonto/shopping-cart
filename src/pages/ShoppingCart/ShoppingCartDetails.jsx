import styled from "styled-components";

const StyledShoppingCartDetails = styled.div`
  background: white;
  display: flex;
  flex-direction: column;

  padding: 12px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  border-radius: 4px;
`;
const ShoppingCartItem = styled(StyledShoppingCartDetails)`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 22px;
  margin: 4px;
`;

const Image = styled.img`
  height: 60px;
`;

const Qty = styled.input`
  width: 35px;
  height: 25px;
`;

const RedButton = styled.button`
  background-color: #ff4d4d;
  color: white;
  padding: 8px 12px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.3s ease;

  &:hover {
    background-color: #ff3333;
  }
`;

const ShoppingCartDetails = ({
  shoppingCart,
  removeFromCart,
  updateQuantity,
}) => {
  const handleQuantityChange = (id, newQuantity) => {
    updateQuantity(id, newQuantity);
  };

  return (
    <StyledShoppingCartDetails>
      {shoppingCart.map((cartItem) => (
        <ShoppingCartItem key={cartItem.details.id}>
          <Image src={cartItem.details.image} alt="alt" />
          <p>{cartItem.details.title}</p>
          <p>
            <b>{cartItem.details.price}â‚¬</b>
          </p>
          <Qty
            type="number"
            min={1}
            max={10}
            step={1}
            value={cartItem.quantity}
            onChange={(e) =>
              handleQuantityChange(cartItem.details.id, e.target.value)
            }
          />
          <RedButton onClick={() => removeFromCart(cartItem.details.id)}>
            Remove
          </RedButton>
        </ShoppingCartItem>
      ))}
    </StyledShoppingCartDetails>
  );
};

export default ShoppingCartDetails;
