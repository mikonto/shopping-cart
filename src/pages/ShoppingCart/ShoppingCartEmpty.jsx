import styled from "styled-components";
import { Link } from "react-router-dom";

const StyledShoppingCartEmpty = styled.div`
  background: white;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 12px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  border-radius: 4px;
`;

const BlueButton = styled.button`
  background-color: skyblue;
  padding: 8px 12px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.3s ease;

  &:hover {
    background-color: #6ca5bc;
  }
`;

const Spacer = styled.div`
  height: 15px;
`;

const Title = styled.h1`
  background: #f6f6f6;
  padding: 12px;
  margin: 8px;
`;

const ShoppingCartEmpty = () => {
  return (
    <StyledShoppingCartEmpty>
      <p>Your shopping cart is empty</p>
      <Link to="/products">
        <BlueButton>Shop now</BlueButton>
      </Link>
      <Spacer />
    </StyledShoppingCartEmpty>
  );
};

export default ShoppingCartEmpty;
