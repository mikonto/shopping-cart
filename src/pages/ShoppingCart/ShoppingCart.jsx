import styled from "styled-components";
import ShoppingCartDetails from "./ShoppingCartDetails";
import ShoppingCartCheckout from "./ShoppingCartCheckout";
import ShoppingCartEmpty from "./ShoppingCartEmpty";
import PropTypes from "prop-types";

const StyledShoppingCart = styled.div`
  display: grid;
  gap: 12px;
  flex-grow: 1;
  padding: 12px;
  grid-template-columns: ${({ hasItems, ...props }) => {
    return hasItems ? "repeat(2, 1fr)" : "repeat(1, 1fr)";
  }};

  grid-template-rows: 0.5fr auto;
  justify-content: center;

  @media (max-width: 1200px) {
    grid-template-columns: repeat(1, 1fr);
  }
`;

const Title = styled.h1`
  background: #f6f6f6;
  padding: 12px;
  margin: 8px;
`;

const ShoppingCart = ({ shoppingCart, removeFromCart, updateQuantity }) => {
  const hasItems = shoppingCart.length > 0;

  return (
    <>
      <Title>Shopping cart</Title>
      <StyledShoppingCart style={{ gridTemplateColumns: hasItems ? "repeat(2, 1fr)" : "repeat(1, 1fr)" }}>
        {hasItems ? (
          <>
            <ShoppingCartDetails
              shoppingCart={shoppingCart}
              removeFromCart={removeFromCart}
              updateQuantity={updateQuantity}
            />
            <ShoppingCartCheckout shoppingCart={shoppingCart} />
          </>
        ) : (
          <ShoppingCartEmpty />
        )}
      </StyledShoppingCart>
    </>
  );
};


ShoppingCart.propTypes = {
  shoppingCart: PropTypes.array,
  removeFromCart: PropTypes.func,
  updateQuantity: PropTypes.func,
};

export default ShoppingCart;
