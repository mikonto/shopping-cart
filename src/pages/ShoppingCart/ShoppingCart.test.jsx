import { render, screen } from "@testing-library/react";
import ShoppingCart from "./ShoppingCart";
import { vi } from "vitest";
import { MemoryRouter } from "react-router-dom";

describe("ShoppingCart component", () => {
  const mockShoppingCart = [
    {
      details: {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: "{count: 259, rate: 4.1}",
      },
      quantity: 1,
    },
  ];

  it("renders the correct title", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={mockShoppingCart}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const title = screen.getByRole("heading", { name: /shopping cart/i });
    expect(title).toBeInTheDocument();
  });

  it("renders the correct text", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={[]}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const emptyMessage = screen.getByText("Your shopping cart is empty");
    expect(emptyMessage).toBeInTheDocument();
  });

  it("renders the button in empty shopping cart", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={[]}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole("button", { name: "Shop now" });
    expect(button).toBeInTheDocument();

    const link = screen.getByRole("link", { name: "Shop now" });
    expect(link).toHaveAttribute("href", "/products");
  });

  it("renders the shopping cart details", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={mockShoppingCart}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const expectedTitle = mockShoppingCart[0].details.title;
    expect(screen.getByText(expectedTitle)).toBeInTheDocument();

    const button = screen.getByRole("button", { name: "Remove" });
    expect(button).toBeInTheDocument();
  });

  it("renders the order summary", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={mockShoppingCart}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const emptyMessage = screen.getByText("Order summary");
    expect(emptyMessage).toBeInTheDocument();
  });

  it("renders the button in order summary", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();

    render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={mockShoppingCart}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole("button", { name: "To checkout" });
    expect(button).toBeInTheDocument();
  });

  it("it renders correctly", () => {
    const mockRemoveFromCart = vi.fn();
    const mockUpdateQuantity = vi.fn();
    const { container } = render(
      <MemoryRouter>
        <ShoppingCart
          shoppingCart={mockShoppingCart}
          removeFromCart={mockRemoveFromCart}
          updateQuantity={mockUpdateQuantity}
        />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
