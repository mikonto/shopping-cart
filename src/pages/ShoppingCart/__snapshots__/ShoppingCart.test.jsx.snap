// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingCart component > it renders correctly 1`] = `
<div>
  <h1
    class="sc-ksBlXE QsXHM"
  >
    Shopping cart
  </h1>
  <div
    class="sc-bqWyeZ kFghZs"
    style="grid-template-columns: repeat(2, 1fr);"
  >
    <div
      class="sc-bcXGCL kSAnLr"
    >
      <div
        class="sc-bcXGCL sc-gswONk kSAnLr cVlEDe"
      >
        <img
          alt="alt"
          class="sc-dkrGBB dceXiZ"
          src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
        />
        <p>
          Mens Casual Premium Slim Fit T-Shirts
        </p>
        <p>
          <b>
            22.3
            €
          </b>
        </p>
        <input
          class="sc-hLBatu kGSFQL"
          max="10"
          min="1"
          step="1"
          type="number"
          value="1"
        />
        <button
          class="sc-eDvShL gYxOWQ"
        >
          Remove
        </button>
      </div>
    </div>
    <div
      class="sc-jSVask dYJdOV"
    >
      <p>
        <b>
          Order summary
        </b>
      </p>
      <ul
        class="sc-gKPSgB iTUalq"
      >
        <li
          class="sc-iBYPxa ffchxx"
        >
          1
           x 
          Mens Casual Premium Slim Fit T-Shirts
        </li>
      </ul>
      <div
        class="sc-pyeOL bhQZRJ"
      />
      <ul
        class="sc-gKPSgB iTUalq"
      >
        <li
          class="sc-ftTHlr gqLuly"
        >
          Products total: 
          22.3
          €
        </li>
        <li
          class="sc-ftTHlr gqLuly"
        >
          Shipping costs: 8€
        </li>
      </ul>
      <div
        class="sc-pyeOL bhQZRJ"
      />
      <ul
        class="sc-gKPSgB iTUalq"
      >
        <li
          class="sc-ftTHlr gqLuly"
        >
          <b>
            Total: 
            30.3
            €
          </b>
        </li>
      </ul>
      <button
        class="sc-jrcSHq kJUJDY"
      >
        To checkout
      </button>
    </div>
  </div>
</div>
`;
